### PR TITLE
Valkyrize Knapsack + Review knapsacking effort

### PR DESCRIPTION
**Pull request by jeremyf**
This is the original branch for our Knapsacking effort. 
After getting a bit unruly, we separated and merged all the bits before we started valkyrizing. 
This PR should contain knapsacking + valkyrizing efforts only

ref: 
- https://github.com/scientist-softserv/palni_palci_knapsack/pull/15

Issues: 
- https://github.com/scientist-softserv/palni_palci_knapsack/issues/2
- https://github.com/scientist-softserv/palni_palci_knapsack/issues/12

# TODO:

- [ ] Actors
  - [ ] OerActor -> Transaction [ create as a separate ticket] 
- [x] Controller
- [x] Factories
- [x] Forms (most of this should be handled by the yaml) [Kirk]
  - [x] EtdForm
  - [x] CdlForm
  - [x] OerForm
  - [x] GenericWorkFormDecorator
  - [x] ImageDecorator
  - [x] FileSet
- [x] Helpers
  - [x] Hyrax::OverrideHelperBehavior [SHANA]
- [x] Indexers (most of this should be handled by the yaml) [Shana/Kirk]
  - [x] EtdResourceIndexer
  - [x] CdlResourceIndexer
  - [x] OerResourceIndexer
  - [x] ImageResourceIndexer
  - [x] GenericResourceIndexer
  - [x] FileSet
- [x] Jobs [Kirk]
  - [x] app/jobs/iiif_print/create_relationships_job_decorator.rb
  - [x] app/jobs/attach_files_to_work_job_decorator.rb
    - This seems to be the old implementation to allow pdfs to be displayed in the viewer by converting them to jpgs.
  - [x] app/jobs/create_group_and_add_members_job.rb
  - [x] app/jobs/bulkrax/related_memberships_job.rb
- [x] Models [Kirk]
  - [x] EtdResource
  - [x] CdlResource
  - [x] OerResource
  - [x] GenericWorkDecorator
  - [x] ImageDecorator
  - [x] GenericWorkResourceDecorator (probably just need to alter the yaml)
  - [x] ImageResourceDecorator (probably just need to alter the yaml)
  - [x] FileSet
- [x] Parsers
  - [x] app/parsers/bulkrax/csv_parser_decorator.rb
  - [x] app/parsers/bulkrax/oer_csv_parser.rb
- [x] Presenters (do we need presenters or do we need to handle it a separate way?) => I think we can use the pre-existing presenters (SHANA)
  - [x] CdlResourcePresenter
  - [x] EtdResourcePresenter
  - [x] OerResourcePresenter
- [x] Video Embed form field
- [x] Convert model/controller/etc callbacks to listeners
  - [x] CDL
    - [x] before_destroy :destroy_split_pages (should be handled by IIIF Print)
    - [x] after_destroy :destroy_cdl_group
   - [x]  COLLECTION
     - [x] :remote_featured
- [ ] SPECS (handle in another ticket)
- [ ] LINT (handle in another ticket)

## 🚧 Add files from PALs Hyku not in other repos

3595d6febba77918a0e5e6497aa58aaefb8e7d4d

This is a first pass of the PALs work.  There was a lot of head
scratching and copying things into Hyku Prime.  Together this is a
rough process.

## 🚧 Partial port of PALs Hyku entries

fac9882bdfae73708de08f9c604f7decb747c7d8

## Move functionality to Hyku prime

c0a37d5fb63017697f1ce9fce4a1230f84220786

## Review and override file_set_helper.rb

a274a65e8915fb92d73b4cd6ab7edac56f5f3cf7

## Favor pushing views down into Hyku Prime

0a34b070824a4249ead1adc0dd3f5caad30e033e

## Ongoing review of files that were different in 3 or more repositories

1357812572061321282c422f5ef011b4a12832b6

## Blank out rendering of Generic Work title

63b6b6bf668376264f8001652f2b043a11dc0f31

## Adding collections/show as this is unique to PALs

183fdd706066024b2c0f604222fdde7beb8cf4c7

## Add custom catalog controller for PALs

de326b42d985c52e6a7e869d1f350f9cef7bbc9a

## 🎁 Decorate ApplicationController based on PALs criteria

b3e8bda2d728e2e0a288c55c6bb745f091571299

### CODE REVIEW

File: app/factories/bulkrax/related_member_object_factory.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/factories/bulkrax/related_member_object_factory.rb:1]: Should this go back to Hyku?  It's unclear.

File: app/helpers/blacklight/advanced_search_helper.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module Blacklight
```
Comment (jeremyf) [inline app/helpers/blacklight/advanced_search_helper.rb:3]: Should these go back to Hyku?

File: app/helpers/blacklight/facets_helper_behavior.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+# OVERRIDE Blacklight 6.7.0 to enable markdown in the facets
```
Comment (jeremyf) [inline app/helpers/blacklight/facets_helper_behavior.rb:3]: Should this go back to Hyku?

File: app/helpers/blacklight/catalog_helper_behavior_decorator.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# OVERRIDE blacklight 6 to allow full url for show links for shared search thumbnail
+module Blacklight
+  module CatalogHelperBehaviorDecorator
```
Comment (jeremyf) [inline app/helpers/blacklight/catalog_helper_behavior_decorator.rb:5]: Should this go back to Hyku?  I believe so given that it has `generic_work_url`

File: app/helpers/blacklight/render_constraints_helper_behavior.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/helpers/blacklight/render_constraints_helper_behavior.rb:1]: This goes back to Hyku

File: app/helpers/hyrax/citations_behaviors/common_behavior.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Hyrax
```
Comment (jeremyf) [inline app/helpers/hyrax/citations_behaviors/common_behavior.rb:3]: The citation formatters might go back to Hyku

File: app/helpers/hyrax/citations_behaviors/formatters/chicago_formatter.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# Hyrax Override: Improve Citations Format
```
Comment (jeremyf) [inline app/helpers/hyrax/citations_behaviors/formatters/chicago_formatter.rb:1]: The citation formatters might go back to Hyku

File: app/helpers/hyrax/citations_behaviors/publication_behavior.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Hyrax Override: Improve Citations Format
```
Comment (jeremyf) [inline app/helpers/hyrax/citations_behaviors/publication_behavior.rb:1]: The citation formatters might go back to Hyku

File: app/helpers/hyrax/override_helper_behavior.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/helpers/hyrax/override_helper_behavior.rb:1]: Back to Hyku?

File: app/indexers/app_indexer_decorator.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+module AppIndexerDecorator
```
Comment (jeremyf) [inline app/indexers/app_indexer_decorator.rb:3]: Back to Hyku?

File: app/jobs/attach_files_to_work_job_decorator.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+# OVERRIDE: Hyrax 5.0.1 to handed is_derived
+class AttachFilesToWorkJobDecorator
+  def attach_work(user, work, work_attributes, work_permissions, uploaded_file)
```
Comment (jeremyf) [inline app/jobs/attach_files_to_work_job_decorator.rb:5]: Does this go back to Hyku?

File: app/jobs/bulkrax/related_memberships_job.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/jobs/bulkrax/related_memberships_job.rb:1]: Does this go back to Hyku?

File: app/models/generic_work_decorator.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module GenericWorkDecorator
```
Comment (jeremyf) [inline app/models/generic_work_decorator.rb:3]: Back to Hyku?

File: app/models/concerns/order_metadata_values.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/models/concerns/order_metadata_values.rb:1]: Back to Hyku?

File: app/parsers/bulkrax/csv_parser_decorator.rb [hunk 15]
```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
```
Comment (jeremyf) [inline app/parsers/bulkrax/csv_parser_decorator.rb:1]: Likely something that should go back to Bulkrax

File: app/services/hyrax/accessibility_features_service.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
+module Hyrax
+  module AccessibilityFeaturesService
```
Comment (jeremyf) [inline app/services/hyrax/accessibility_features_service.rb:2]: Are these something to go back to Hyku?

File: app/services/hyrax/accessibility_hazards_service.rb [hunk 17]
```diff
@@ -0,0 +1,25 @@
+module Hyrax
+  module AccessibilityHazardsService
```
Comment (jeremyf) [inline app/services/hyrax/accessibility_hazards_service.rb:2]: Are these something that goes back to Hyku?

File: app/services/hyrax/audience_service.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
+module Hyrax
+  module AudienceService
```
Comment (jeremyf) [inline app/services/hyrax/audience_service.rb:2]: Are these something to send back to Hyku?

File: app/services/hyrax/contributing_library_service.rb [hunk 19]
```diff
@@ -0,0 +1,20 @@
+module Hyrax
+  # Provide select options for the types field
+  class ContributingLibraryService < QaSelectService
```
Comment (jeremyf) [inline app/services/hyrax/contributing_library_service.rb:3]: Is this something to send back to Hyku?

File: app/services/hyrax/learning_resource_types_service.rb [hunk 20]
```diff
@@ -0,0 +1,25 @@
+module Hyrax
+  module LearningResourceTypesService
```
Comment (jeremyf) [inline app/services/hyrax/learning_resource_types_service.rb:2]: Something that belongs in Hyku?

File: app/services/hyrax/oer_types_service.rb [hunk 21]
```diff
@@ -0,0 +1,25 @@
+module Hyrax
+  module OerTypesService
```
Comment (jeremyf) [inline app/services/hyrax/oer_types_service.rb:2]: Something that belongs in Hyku?

File: app/views/hyrax/admin/workflows/index.html.erb [hunk 22]
```diff
@@ -0,0 +1,73 @@
+<%# OVERRIDE: Hyrax v5.0.1 to show the user/depositor's display_name if available %>
```
Comment (jeremyf) [inline app/views/hyrax/admin/workflows/index.html.erb:1]: Perhaps something for Hyku?

File: app/views/hyrax/collections/show.html.erb [hunk 23]
```diff
@@ -0,0 +1,141 @@
+<%# OVERRIDE Hyrax 5.0.1 and Hyku 6.0 to update the layout of the collection show page to be more visually appealing %>
```
Comment (jeremyf) [inline app/views/hyrax/collections/show.html.erb:1]: We may be able to apply this file back into Hyku.  But that is a matter of UI verification.

File: lib/active_job_tenant_decorator.rb [hunk 24]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module ActiveJobTenant
```
Comment (jeremyf) [inline lib/active_job_tenant_decorator.rb:3]: We may need to move some of this out.  But a word of caution, Raven is specific to PALs and not generalizable to Hyku.

File: lib/oai/provider/model_decorator.rb [hunk 25]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
```
Comment (jeremyf) [inline lib/oai/provider/model_decorator.rb:2]: Do these go back to Hyku?

File: lib/omni_auth/strategies/open_id_connect_decorator.rb [hunk 26]
```diff
@@ -0,0 +1,143 @@
+module OmniAuth
```
Comment (jeremyf) [inline lib/omni_auth/strategies/open_id_connect_decorator.rb:1]: The work authorization is specific to ReShare but other aspects might be relevant to Hyku.  Let's hold on moving this back to Hyku.

**Comment by ShanaLMoore** [comment]
TODO: Cut hyku release. update this branch. solve build issue

**Pull request merged**